File: app/models/subject.rb
class Subject < ActiveRecord::Base
  # entirely new model
  # each department has many subjects and there are many courses within a subject
  belongs_to :department
  has_many   :courses, dependent: :destroy
  # subject must have a full name and abbrevation (longname and shortname)
  validates  :shortname, presence: true, uniqueness: true
  validates  :longname, presence: true, uniqueness: true
  default_scope { order(shortname: :asc) }
end

File: packages/yacs-api-client/src/lib/subject.ts
import {
  Model,
  SpraypaintBase,
  Attr,
  BelongsTo,
  HasMany
} from "spraypaint";

import { ApplicationRecord } from "./application-record";
import { School } from "./school";
import { Course } from "./course";

@Model()
export class Subject extends ApplicationRecord {
	static jsonapiType = "subjects";

	@Attr() longname: string;
	@Attr() shortname: string;
	@Attr() uuid: string;

	@BelongsTo() school: School;
	@HasMany() courses: Course[];
}
